distribution/xfer/transfer.go

```diff
@@ -155,18 +154,23 @@ func (t *transfer) Watch(progressOutput progress.Output) *Watcher {
 		defer func() {
 			close(w.running)
 		}()
-		done := false
+		var (
+			done           bool
+			lastWritten    progress.Progress
+			hasLastWritten bool
+		)
 		for {
 			t.mu.Lock()
 			hasLastProgress := t.hasLastProgress
 			lastProgress := t.lastProgress
 			t.mu.Unlock()
 
-			// This might write the last progress item a
-			// second time (since channel closure also gets
-			// us here), but that's fine.
-			if hasLastProgress {
+			// Make sure we don't write the last progress item
+			// twice.
+			if hasLastProgress && (!done || !hasLastWritten || lastProgress != lastWritten) {
```

Comment:
Changing it to:
```
if hasLastProgress && (lastWritten != lastProgress) {
    progressOutput.WriteProgress(lastProgress)
    lastWritten = lastProgress
}
```
does appear to work - at least for the busybox case.

I agree with @calavera, some unit testcases (if possible) might be good to verify we didn't break anything.